ddtrace/tracer/sampler.go

```diff
@@ -158,49 +158,115 @@ func (ps *prioritySampler) apply(spn *span) {
-// These rules can match based on the span's Service, Name or both.
-// When making a sampling decision, the rules are checked in order until
-// a match is found.
-// If a match is found, the rate from that rule is used.
-// If no match is found, and the DD_TRACE_SAMPLE_RATE environment variable
-// was set to a valid rate, that value is used.
-// Otherwise, the rules sampler didn't apply to the span, and the decision
-// is passed to the priority sampler.
-//
-// The rate is used to determine if the span should be sampled, but an upper
-// limit can be defined using the DD_TRACE_RATE_LIMIT environment variable.
-// Its value is the number of spans to sample per second.
-// Spans that matched the rules but exceeded the rate limit are not sampled.
+// rulesSampler holds instances of trace sampler and single span sampler, that are configured with the given set of rules.
+// traceRulesSampler samples trace spans based on a user-defined set of rules and might impact sampling decision of the trace.
+// singleSpanRulesSampler samples individual spans based on a separate user-defined set of rules and
+// cannot impact the trace sampling decision.
 type rulesSampler struct {
-	rules      []SamplingRule // the rules to match spans with
-	globalRate float64        // a rate to apply when no rules match a span
-	limiter    *rateLimiter   // used to limit the volume of spans sampled
+	traceRulesSampler      *traceRulesSampler
+	singleSpanRulesSampler *singleSpanRulesSampler
 }
 
 // newRulesSampler configures a *rulesSampler instance using the given set of rules.
+// Rules are split between trace and single span sampling rules according to their type.
+// Such rules are user-defined through environment variable or WithSamplingRules option.
 // Invalid rules or environment variable values are tolerated, by logging warnings and then ignoring them.
 func newRulesSampler(rules []SamplingRule) *rulesSampler {
+	var spanRules, traceRules []SamplingRule
+	for _, rule := range rules {
+		if rule.Type == SamplingRuleSingleSpan {
+			rule.limiter = newSingleSpanRateLimiter(rule.MaxPerSecond)
+			spanRules = append(spanRules, rule)
+		} else {
+			traceRules = append(traceRules, rule)
+		}
+	}
 	return &rulesSampler{
-		rules:      rules,
-		globalRate: globalSampleRate(),
-		limiter:    newRateLimiter(),
+		traceRulesSampler:      newTraceRulesSampler(traceRules),
+		singleSpanRulesSampler: newSingleSpanRulesSampler(spanRules),
 	}
 }
 
-// samplingRulesFromEnv parses sampling rules from the DD_TRACE_SAMPLING_RULES
-// environment variable.
+const (
+	// SamplingRuleTrace specifies that if a span matches a sampling rule of this type,
+	// an entire trace might be kept based on the given span and sent to the agent.
+	SamplingRuleTrace = iota
+	// SamplingRuleSingleSpan specifies that if a span matches a sampling rule,
+	// an individual span might be kept and sent to the agent.
+	SamplingRuleSingleSpan
```

Comment:
Since we don't have a `SamplingRuleMultipleSpan`, why not name this `SamplingRuleSpan`? Btw. this is the right place to explain what this type of sampling rule does and how it affects things, in as much detail as needed.